File: src/main/kotlin/com/marcel/baumgardt/repository/ShowingRepository.java
package com.marcel.baumgardt.repository;

import com.marcel.baumgardt.model.db.Showing;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
public interface ShowingRepository extends JpaRepository<Showing, Long> {

    @Modifying(flushAutomatically = true, clearAutomatically = true)
    @Query(nativeQuery = true,
            value = "UPDATE sho " +
                    "SET sho.sho_price = ?1" +
                    "FROM dbo.showings WITH(index(sho_pk), nolock) " +
                    "WHERE sho.sho_id in (?2) ")
    int updatePriceOfShowingsByIds(Double price, List<Long> showingIds);

    @Modifying(flushAutomatically = true, clearAutomatically = true)
    @Query(nativeQuery = true,
            value = "UPDATE sho " +
                    "SET sho.sho_price = ?1" +
                    "FROM dbo.showings WITH(index(sho_cin_mov_id), nolock) " +
                    "WHERE sho.sho_cin_id = ?2) " +
                    "AND sho.sho_mov_id = ?3")
    int updatePriceOfShowingsByCinemaIdAndMovieId(Double price, Long cinemaId, Long movieId);

    @Query(nativeQuery = true,
            value = "SELECT sho " +
                    "FROM dbo.showings WITH(index(sho_cin_mov_id), nolock) " +
                    "WHERE sho.sho_cin_id = ?1) " +
                    "AND sho.sho_mov_id = ?2")
    List<Showing> getShowingsByCinemaIdAndMovieId(Long cinemaId, Long movieId);

    @Query(nativeQuery = true,
            value = "SELECT count(sho) " +
                    "FROM dbo.showings WITH(index(sho_cin_mov_id), nolock) " +
                    "WHERE sho.sho_cin_id = ?1) " +
                    "AND sho.sho_mov_id = ?2")
    int countShowingsByCinemaIdAndMovieId(Long cinemaId, Long movieId);

    @Modifying(flushAutomatically = true, clearAutomatically = true)
    @Query(nativeQuery = true,
            value = "DELETE sho " +
                    "FROM dbo.showings WITH(index(sho_cin_mov_id), nolock) " +
                    "WHERE sho.sho_cin_id = ?1) " +
                    "AND sho.sho_mov_id = ?2")
    int deleteAllByCinemaIdAndMovieId(Long cinemaId, Long movieId);
}

File: src/main/kotlin/com/marcel/baumgardt/repository/ShowingRepository.kt
package com.marcel.baumgardt.repository

import com.marcel.baumgardt.model.db.Showing
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Modifying
import org.springframework.data.jpa.repository.Query
import org.springframework.stereotype.Repository
import java.time.OffsetTime
import javax.transaction.Transactional

@Repository
interface ShowingRepository : JpaRepository<Showing, Long> {

    @Transactional
    @Modifying(flushAutomatically = true, clearAutomatically = true)
    @Query(
        nativeQuery = true, value = "UPDATE dbo.showings " +
                "SET sho_price = ?1 " +
                "WHERE sho_id in (?2)"
    )
    fun updatePriceOfShowingsByIds(price: Double, showingIds: List<Long>): Int

    @Transactional
    @Modifying(flushAutomatically = true, clearAutomatically = true)
    @Query(
        nativeQuery = true, value = "UPDATE dbo.showings " +
                "SET sho_price = ?1 " +
                "WHERE sho_cin_id = ?2 " +
                "AND sho_mov_id = ?3"
    )
    fun updatePriceOfShowingsByCinemaIdAndMovieId(price: Double, cinemaId: Long, movieId: Long): Int


    @Query(
        nativeQuery = true, value = "SELECT * " +
                "FROM dbo.showings sho " +
                "LEFT JOIN dbo.cinemas cin on cin.cin_id = sho.sho_cin_id " +
                "LEFT JOIN dbo.movies mov on mov.mov_id = sho.sho_mov_id " +
                "WHERE sho.sho_cin_id = ?1 " +
                "AND sho.sho_mov_id = ?2"
    )
    fun getShowingsByCinemaIdAndMovieId(cinemaId: Long, movieId: Long): List<Showing>

    @Query(
        nativeQuery = true, value = "SELECT * " +
                "FROM dbo.showings sho " +
                "WHERE sho.sho_cin_id = ?1 " +
                "AND sho.sho_mov_id = ?2 " +
                "AND sho.sho_day = ?3 " +
                "AND sho.sho_time in (?4)"
    )
    fun getShowingsByCinemaIdAndMovieIdFromDayAndTime(
        cinemaId: Long,
        movieId: Long,
        day: Int,
        times: List<OffsetTime>
    ): List<Showing>

    @Query(
        nativeQuery = true, value = "SELECT count(*) " +
                "FROM dbo.showings sho " +
                "WHERE sho.sho_cin_id = ?1 " +
                "AND sho.sho_mov_id = ?2"
    )
    fun countShowingsByCinemaIdAndMovieId(cinemaId: Long, movieId: Long): Int

    @Transactional
    @Modifying(flushAutomatically = true, clearAutomatically = true)
    @Query(
        nativeQuery = true, value = "DELETE " +
                "FROM dbo.showings sho " +
                "WHERE sho.sho_cin_id = ?1 " +
                "AND sho.sho_mov_id = ?2"
    )
    fun deleteAllByCinemaIdAndMovieId(cinemaId: Long, movieId: Long): Int
}